### PR TITLE
udev: net_id: introduce predictable names for netdevsim

### PR DESCRIPTION
**Pull request by jpirko**
In order to properly and predictably name netdevsim netdevices,
introduce a separate implementation, as the netdevices reside on a
specific netdevsim bus. Note that this applies only to netdevsim devices
created using sysfs, because those expose phys_port_name attribute.

Signed-off-by: Jiri Pirko <jiri@mellanox.com>

EDIT: This is equivalent to #12328. Please see the previous discussion on that page.

### CODE REVIEW

**Comment by yuwata** [comment]
(Rather than creating new PR, please use force-push on update.)

**Comment by jpirko** [comment]
> (Rather than creating new PR, please use force-push on update.)

Okay, I'm sorry for my poor knowledge of github. What should I do? Btw, the last push failed one check, probably for some internal checker issue, so this pr is no change, I only needed the check to be ran again.

**Comment by yuwata** [comment]
> > (Rather than creating new PR, please use force-push on update.)
> 
> Okay, I'm sorry for my poor knowledge of github. What should I do?

E.g., `git push jpirko netdevsim --force`.

> Btw, the last push failed one check, probably for some internal checker issue, so this pr is no change, I only needed the check to be ran again.

The CIs are restarted when you push new commits on the remote branch. So, when you want to restart CIs without changing commits, then for example,
```
$ git commit --amend
$ git push jpirko netdevsim --force
```
is fine.

**Comment by jpirko** [comment]
> > > (Rather than creating new PR, please use force-push on update.)
> > 
> > 
> > Okay, I'm sorry for my poor knowledge of github. What should I do?
> 
> E.g., `git push jpirko netdevsim --force`.
> 
> > Btw, the last push failed one check, probably for some internal checker issue, so this pr is no change, I only needed the check to be ran again.
> 
> The CIs are restarted when you push new commits on the remote branch. So, when you want to restart CIs without changing commits, then for example,
> 
> ```
> $ git commit --amend
> $ git push jpirko netdevsim --force
> ```
> 
> is fine.

Okay, I got it. Should I close this one and push the original one or is it fine for now? Thanks!

**Comment by yuwata** [comment]
> Okay, I got it. Should I close this one and push the original one or is it fine for now? Thanks!

I am not sure it is possible to reopen the original PR. I think it is OK to keep this one.

**Comment by yuwata** [comment]
From https://github.com/systemd/systemd/pull/12328#issuecomment-483558251,

> I'm currently changing the way netdevsim is instantiated, from "ip link add" creation to using sysfs. More details here: https://www.spinics.net/lists/netdev/msg564284.html

Could you elaborate more? `ip link add` requires an interface name. Why the name needs to be renamed?

**Comment by jpirko** [comment]
> From [#12328 (comment)](https://github.com/systemd/systemd/pull/12328#issuecomment-483558251),
> 
> > I'm currently changing the way netdevsim is instantiated, from "ip link add" creation to using sysfs. More details here: https://www.spinics.net/lists/netdev/msg564284.html
> 
> Could you elaborate more? `ip link add` requires an interface name. Why the name needs to be renamed?

New interface is going to make netdevsim to behave more alike the actual hw. That means the device is going to be added using sysfs and then the same bus device creation and probe is going to happen as if an acrual hw would be added on a bus. During the process, the netdevices are going to be created with default names and udev needs to rename them according to netdevsim bus addr and port name. There will be no longer possible to use "ip link add" to create netdevsim devices.

File: src/udev/udev-builtin-net_id.c [hunk 1]
```diff
@@ -794,6 +797,44 @@ static int names_mac(sd_device *dev, struct netnames *names) {
         return 0;
 }
 
+static int names_netdevsim(sd_device *dev, struct netnames *names) {
+        sd_device *netdevsimdev;
+        const char *sysname;
+        unsigned int addr;
```
Comment (keszybz) [inline src/udev/udev-builtin-net_id.c]: `unsigned int` → `unsigned` please.

File: src/udev/udev-builtin-net_id.c [hunk 2]
```diff
@@ -794,6 +797,44 @@ static int names_mac(sd_device *dev, struct netnames *names) {
...
+        r = sd_device_get_parent_with_subsystem_devtype(dev, "netdevsim", NULL, &netdevsimdev);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_sysname(netdevsimdev, &sysname);
+        if (r < 0)
+                return r;
+
+        if (sscanf(sysname, "netdevsim%u", &addr) != 1)
+                return -EINVAL;
```
Comment (keszybz) [inline src/udev/udev-builtin-net_id.c:822]: I'd prefer
```C
const char *t;
t = startswith(sysname, "netdevsim");
if (!t)
     return -EINVAL;
r = safe_atou(t, &addr);
if (r < 0)
     return log_debug_errno(r, "Failed to parse netdevsim device number %s: %m", t);
```
This does better error handling, in particular trailing garbage is detected properly.

File: src/udev/udev-builtin-net_id.c [hunk 3]
```diff
@@ -31,6 +31,7 @@
  *                                         — USB port number chain
  *   v<slot>                               - VIO slot number (IBM PowerVM)
  *   a<vendor><model>i<instance>           — Platform bus ACPI instance id
+ *   i<addr>n<phys_port_name	           — Netdevsim bus address and port name
```
Comment (keszybz) [inline src/udev/udev-builtin-net_id.c]: Closing bracket is missing. Also, it seems that the `phys_port_name` part is conditional. It should be in square brackets.

File: src/udev/udev-builtin-net_id.c [hunk 4]
```diff
@@ -794,6 +797,44 @@ static int names_mac(sd_device *dev, struct netnames *names) {
...
+
+        (void) sd_device_get_sysattr_value(dev, "phys_port_name", &port_name);
+
+        s = names->netdevsim_path;
+        l = sizeof(names->netdevsim_path);
+        l = strpcpyf(&s, l, "i%u", addr);
+        if (port_name)
+                l = strpcpyf(&s, l, "n%s", port_name);
+        else
+                return 0;
```
Comment (keszybz) [inline src/udev/udev-builtin-net_id.c]: Please don't use the `strpcpyf` functions, they are so hard to parse. What about
```C
    bool ok;
    if (port_name)
          ok = snprintf_ok(names->netdevsim_path, sizeof(names->netdevsim_path), "i%un%s", addr, port_name);
    else
          ok = snprintf_ok(names->netdevsim_path, sizeof(names->netdevsim_path), "i%u", addr);
    if (!ok)
          return -ENOBUFS;
    names->type = NET_NETDEVSIM;
    return 0;
```
This detects truncation properly and sets type always.

File: src/udev/udev-builtin-net_id.c [hunk 5]
```diff
@@ -897,6 +938,20 @@ static int builtin_net_id(sd_device *dev, int argc, char *argv[], bool test) {
                 return 0;
         }
 
+        /* get netdevsim path names */
+        if (names_netdevsim(dev, &names) >= 0 && names.type == NET_NETDEVSIM) {
+                char str[IFNAMSIZ];
+
+                if (names.netdevsim_path[0] &&
```
Comment (keszybz) [inline src/udev/udev-builtin-net_id.c]: Why check `names.netdevsim_path[0]`? `names_netdevsim` does all the checks already.

File: src/udev/udev-builtin-net_id.c [hunk 6]
```diff
@@ -897,6 +938,20 @@ static int builtin_net_id(sd_device *dev, int argc, char *argv[], bool test) {
...
+                    snprintf_ok(str, sizeof str, "%s%s", prefix, names.netdevsim_path))
+                        udev_builtin_add_property(dev, test, "ID_NET_NAME_PATH", str);
+
+                if (names.netdevsim_path[0] &&
```
Comment (keszybz) [inline src/udev/udev-builtin-net_id.c]: Both `if`s are the same, so the same value is used for both ID_NET_NAME_PATH and ID_NET_NAME_SLOT. Is this on purpose? I don't think it makes much sense to provide the same name twice.

File: src/udev/udev-builtin-net_id.c (hunk 3 above)
Comment (yuwata) [inline src/udev/udev-builtin-net_id.c]: missing `>`.

File: src/udev/udev-builtin-net_id.c (hunk 1 above)
Comment (yuwata) [inline src/udev/udev-builtin-net_id.c]: Please use `unsigned` instead of `unsigned int`.

File: src/udev/udev-builtin-net_id.c [hunk 7]
```diff
@@ -794,6 +797,44 @@ static int names_mac(sd_device *dev, struct netnames *names) {
...
+                return r;
+
+        r = sd_device_get_sysname(netdevsimdev, &sysname);
+        if (r < 0)
+                return r;
+
+        if (sscanf(sysname, "netdevsim%u", &addr) != 1)
+                return -EINVAL;
+
+        (void) sd_device_get_sysattr_value(dev, "phys_port_name", &port_name);
```
Comment (yuwata) [inline src/udev/udev-builtin-net_id.c]: The current implementation does not allow devices without `port_name`. So, please return earlier than writing the address in `names->netdevsim_path`. That is,
```
r = sd_device_get_sysattr_value(dev, "phys_port_name", &port_name);
if (r < 0)
        return r;
```

File: src/udev/udev-builtin-net_id.c [hunk 8]
```diff
@@ -897,6 +938,20 @@ static int builtin_net_id(sd_device *dev, int argc, char *argv[], bool test) {
...
+        if (names_netdevsim(dev, &names) >= 0 && names.type == NET_NETDEVSIM) {
+                char str[IFNAMSIZ];
+
+                if (names.netdevsim_path[0] &&
+                    snprintf_ok(str, sizeof str, "%s%s", prefix, names.netdevsim_path))
+                        udev_builtin_add_property(dev, test, "ID_NET_NAME_PATH", str);
+
+                if (names.netdevsim_path[0] &&
+                    snprintf_ok(str, sizeof str, "%s%s", prefix, names.netdevsim_path))
+                        udev_builtin_add_property(dev, test, "ID_NET_NAME_SLOT", str);
```
Comment (yuwata) [inline src/udev/udev-builtin-net_id.c]: I do not think `ID_NET_NAME_SLOT` should be set.

File: src/udev/udev-builtin-net_id.c (hunk 3 above)
Comment (jpirko) [reply src/udev/udev-builtin-net_id.c]: Will add the bracket. The phys_port_name part is not conditional. It is always present.

File: src/udev/udev-builtin-net_id.c (hunk 4 above)
Comment (jpirko) [reply src/udev/udev-builtin-net_id.c]: I don't mind, I just copy&pasted and adjusted existing code. Will do it like you ask.

File: src/udev/udev-builtin-net_id.c (hunk 2 above)
Comment (jpirko) [reply src/udev/udev-builtin-net_id.c:822]: Okay, again, I just copy&pasted and adjusted the existing code. Will do.

File: src/udev/udev-builtin-net_id.c (hunk 6 above)
Comment (jpirko) [reply src/udev/udev-builtin-net_id.c]: Okay. I just followed what is already in the code for NET_USB and NET_BCMA. Will remove ID_NET_NAME_SLOT

File: src/udev/udev-builtin-net_id.c (hunk 2 above)
Comment (jpirko) [reply src/udev/udev-builtin-net_id.c:822]: Hmm, your code tries to read uint from the beginning of the string :/

Comment (jpirko) [reply src/udev/udev-builtin-net_id.c:822]: I don't know, is this really needed? The rest of the code just uses sscanf. The input is sysfs file name, not user input. Looks like overkill to me...

**Comment by jpirko** [comment]
> The implementation looks generally OK, some minor improvements are necessary.
> 
> But the whole feature needs to be conditionalized using a new naming scheme entry. See `src/udev/net/naming-scheme.h`.

So if I understand correct, I need to have a separate commit to add:
NAMING_V243 = NAMING_V240 | NAMING_V239 | NAMING_INFINIBAND | NAMING_ZERO_ACPI_INDEX | NAMING_ALLOW_RERENAMES | NAMING_NETDEVSIM
?

File: src/udev/udev-builtin-net_id.c (hunk 2 above)
Comment (keszybz) [reply src/udev/udev-builtin-net_id.c:822]: I fixed the error in the snippet. I guess sscanf is ok in this particular case.

**Comment by keszybz** [comment]
> NAMING_V243 = NAMING_V240 | NAMING_V239 | NAMING_INFINIBAND | NAMING_ZERO_ACPI_INDEX | NAMING_ALLOW_RERENAMES | NAMING_NETDEVSIM
> ?

`NAMING_V240` already includes all those bits, so you only should add the new one:
`NAMING_V243 = NAMING_V240 | NAMING_INFINIBAND;`.

It doesn't have to be a separate commit. Whatever makes the most sense.

**Comment by jpirko** [comment]
> > NAMING_V243 = NAMING_V240 | NAMING_V239 | NAMING_INFINIBAND | NAMING_ZERO_ACPI_INDEX | NAMING_ALLOW_RERENAMES | NAMING_NETDEVSIM
> > ?
> 
> `NAMING_V240` already includes all those bits, so you only should add the new one:
> `NAMING_V243 = NAMING_V240 | NAMING_INFINIBAND;`.
> 
> It doesn't have to be a separate commit. Whatever makes the most sense.

Got it. Thanks!

**Comment by jpirko** [comment]
I force pushed updated commit, checks are ok, how do I proceed now? There are still "changes requested" but all were addressed and resolved.

**Comment by keszybz** [comment]
>  There are still "changes requested" but all were addressed and resolved.

Please ignore those. They will be updated if/when the new review happens. (But if not all reviewers re-review, then they'll stay like this. A limitation of github.)

**Comment by keszybz** [comment]
LGTM.

**Comment by keszybz** [comment]
This will also need a news entry, but let's prepare that before the release.
